Source/Core/Core/IOS/USB/Emulated/Infinity.cpp

```diff
@@ -0,0 +1,987 @@
+  for (s8 i = 3; i >= 0; i--)
+  {
+    figure_data[15 - i] = u8((checksum >> i * 8) & 0xFF);
+  }
+  DEBUG_LOG_FMT(IOS_USB, "Block 1: \n{}", HexDump(figure_data.data(), figure_data.size()));
+  Common::SHA1::Digest digest = Common::SHA1::CalculateDigest(sha1_calc);
+  // Infinity AES keys are the first 16 bytes of the SHA1 Digest, every set of 4 bytes need to be
+  // reversed due to endianness
+  std::vector<u8> infinity_aes_key = {};
+  for (int i = 0; i < 4; i++)
+  {
+    for (int x = 3; x >= 0; x--)
+    {
+      infinity_aes_key.push_back(digest[x + (i * 4)]);
+    }
+  }
+
+  // Create AES Encrypt context based on AES key, use this to encrypt the character data and 4 blank
+  // blocks
+  std::unique_ptr<Common::AES::Context> ctx =
+      Common::AES::CreateContextEncrypt(infinity_aes_key.data());
+  std::array<u8, 16> encrypted_block = {};
+  std::array<u8, 16> encrypted_blank = {};
+
+  ctx->CryptIvZero(figure_data.data(), encrypted_block.data(), figure_data.size());
+  ctx->CryptIvZero(BLANK_BLOCK.data(), encrypted_blank.data(), BLANK_BLOCK.size());
+
+  memcpy(&file_data[0], uid_data.data(), uid_data.size());
+  memcpy(&file_data[16], encrypted_block.data(), encrypted_block.size());
+  memcpy(&file_data[16 * 0x04], encrypted_blank.data(), encrypted_blank.size());
+  memcpy(&file_data[16 * 0x08], encrypted_blank.data(), encrypted_blank.size());
+  memcpy(&file_data[16 * 0x0C], encrypted_blank.data(), encrypted_blank.size());
+  memcpy(&file_data[16 * 0x0D], encrypted_blank.data(), encrypted_blank.size());
+
+  DEBUG_LOG_FMT(IOS_USB, "File Data: \n{}", HexDump(file_data.data(), file_data.size()));
+
+  inf_file.WriteBytes(file_data.data(), file_data.size());
+
+  return true;
+}
+
+const std::map<const char*, const InfinityFigureInfo>& InfinityBase::GetFigureList()
+{
+  return list_infinity_figures;
+}
+
+std::string InfinityBase::FindFigure(u16 number) const
+{
+  for (auto it = list_infinity_figures.begin(); it != list_infinity_figures.end(); it++)
+  {
+    if (it->second.figure_number == number)
+    {
+      return it->first;
+    }
+  }
+  return "";
+}
+
+void InfinityBase::GenerateSeed(u32 seed)
+{
+  m_random_a = 0xF1EA5EED;
+  m_random_b = seed;
+  m_random_c = seed;
+  m_random_d = seed;
+
+  for (int i = 0; i < 23; i++)
+  {
+    GetNext();
+  }
+}
+
+u32 InfinityBase::GetNext()
+{
+  u32 a = m_random_a;
+  u32 b = m_random_b;
+  u32 c = m_random_c;
+  u32 ret = std::rotl(m_random_b, 27);
+
+  u32 temp = (a + ((ret ^ 0xFFFFFFFF) + 1)) & 0xFFFFFFFF;
+  b = (b ^ std::rotl(c, 17)) & 0xFFFFFFFF;
+  a = m_random_d;
+  c = (c + a) & 0xFFFFFFFF;
+  ret = (b + temp) & 0xFFFFFFFF;
+  a = (a + temp) & 0xFFFFFFFF;
+
+  m_random_c = a;
+  m_random_a = b;
+  m_random_b = c;
+  m_random_d = ret;
+
+  return ret;
+}
+
+u64 InfinityBase::Scramble(u32 num_to_scramble, u32 garbage)
+{
+  u64 mask = 0x8E55AA1B3999E8AA;
+  u64 ret = 0;
+
+  for (int i = 0; i < 64; i++)
+  {
+    ret = ret << 1;
+
+    if ((mask & 0x01) > 0)
```

Comment:
This works, but is a somewhat unusual way of checking if the bit is set. I'd write the check like this instead:
```suggestion
    if ((mask & 1) != 0)
```